rfd/xxxx-bot-for-cert-renewals.md

```diff
@@ -0,0 +1,627 @@
+   the certificate gains access to the system for a fairly long period of time
+   (and malicious users would of course do so immediately). One might speculate
+   that a malicious user would not need a full hour to spawn a remote shell on
+   every node the bot user can access.
+
+   Presumably, legitimate users should notice that their certificate bot has
+   failed to renew (this should be a very loud error). If they accidentally
+   started the bot twice, there's at least a 50% chance that no further
+   action is required on their end to resolve the issue (if the correct bot
+   wins or if the start ordering didn't matter).
+
+Another traceability enhancement would be to ensure renewable certificates
+have a unique serial number. Upon renewal, we can embed the previous
+certificate serial as another metadata / extension field. Should a certificate
+be compromised, it will then be possible to trace the the tree of certificate
+renewals. (*TODO (Tim): I suppose in practice the generation counter ensures
+a flat hierarchy, so this may not be useful.*)
+
+Lastly, regardless of the initial certificate's TTL, we can have the auth
+server ensure that, when renewing, renewed certificate TTLs may only decrease
+in length. This would prevent an attacker from stealing a renewable certificate
+and immediately requesting a new certificate with the auth server's max
+renewable TTL.
+
+#### API Client Refresh
+
+As of now, our [API client](./0010-api.md) is initialized with a set of TLS
+credentials and expects those credentials be valid for the lifetime of the
+client.
+
+In order to continue communicating with the cluster, the bot will need to
+inform the API client of the new credentials after a renewal. As part of this
+effort, the client will be updated to allow for refreshing itself. This process
+initializes a new client that attempts to connect with the new credentials, and
+closes the original client when the new client is successfully connected with
+the cluster.
+
+```go
+func (c *Client) Refresh(ctx context.Context) error {
+    // use the existing config to generate a new client
+    newClient, err := connect(ctx, c.c)
+
+    c.Close() // close the original client
+
+    *c = newClient
+}
+```
+
+It will be the responsibility of the caller of `Refresh` to reinitialize any
+watches or streams that the client may have been running prior to the refresh.
+
+### Certificate Specification
+
+The bot will be configured with one or more certificate specs, which can be
+supplied on the command line in the format of:
+
+```
+--cert=DESTINATION,RELOAD
+```
+
+...or via a `tbot.yaml` configuration file.
+
+*TODO*: how can a user specify DNS names to include in the certs?
+
+*TODO* (Tim): Non-filesystem destinations are likely to be very difficult to
+encode as CLI flags (e.g. k8s secrets). Perhaps the CLI should only support
+local FS and other sinks should defer to `tbot.yaml`?
```

Comment:
Could we avoid the need for tbot.yaml entirely if we made the flag only work for filesystem sinks, and then add a config file in the future if/when we decide to support other options?

If we can keep the CLI simple enough, I like the idea of not needing another config file.

Edit: should have read further. Looks like you're proposing exactly this.

Reply:
Yup. This is itself sort of a second-order effect stemming from moving multiple certs to a YAML file. When we do that, it makes sense to then also move "advanced" destinations to the file too (in my opinion, at least).